### PR TITLE
store/tikv: Update client-go to fix the issue that NotLeader error might lead to unnecessary backoff

### LINKED ISSUE #50418: TiKV reporting NotLeader error with new leader info may lead to unnecessary backoff
## Bug Report

Please refer to https://github.com/tikv/client-go/issues/1112 for detail about the problem.

This is also possibly the cause of #49679

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Issue Number: close #50418 

Problem Summary:

### What changed and how does it work?

Updated client-go to include https://github.com/tikv/client-go/pull/1115 which fixes the issue that leader changing sometimes may cause unnecessary backoff.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix an issue that leader changing sometimes may cause unnecessary backoff
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -105,7 +105,7 @@ require (
 	github.com/stretchr/testify v1.8.4
 	github.com/tdakkota/asciicheck v0.2.0
 	github.com/tiancaiamao/appdash v0.0.0-20181126055449-889f96f722a2
-	github.com/tikv/client-go/v2 v2.0.8-0.20231227070846-61c486af13a5
+	github.com/tikv/client-go/v2 v2.0.8-0.20240116040509-1fd589dbf277
```
Comment (cfzjywxk) [inline go.mod:108]: The bazel information is not updated.

**Review by cfzjywxk — APPROVED** [review]

**Review by zyguan — APPROVED** [review]

**Review by crazycs520 — APPROVED** [review]

**Comment by MyonKeminta** [comment]
`TestIndexJoin31494` fails stably in this PR. Not yet figured out the reason.

```
    executor_issue_test.go:320: 
        	Error Trace:	pkg/executor/test/issuetest/executor_issue_test.go:320
        	Error:      	Should be true
        	Test:       	TestIndexJoin31494
```

```go
/* 320 */	require.True(t, exeerrors.ErrMemoryExceedForQuery.Equal(err)) // err: tikv:1317 Query execution was interrupted
```

**Pull request closed without merging**